feature_engine/encoding/similarity_encoder.py

```diff
@@ -0,0 +1,298 @@
+        X = check_X(X)
+        self._check_or_select_variables(X)
+        self._get_feature_names_in(X)
+        self.encoder_dict_ = {}
+
+        if self.handle_missing == 'error':
+            _check_contains_na(X, self.variables_)
+            for var in self.variables_:
+                self.encoder_dict_[var] = (
+                    X[var]
+                    .value_counts()
+                    .head(self.top_categories)
+                    .index
+                    .tolist()
+                )
+        elif self.handle_missing == 'impute':
+            for var in self.variables_:
+                self.encoder_dict_[var] = (
+                    X[var]
+                    .fillna('')
+                    .value_counts()
+                    .head(self.top_categories)
+                    .index
+                    .tolist()
+                )
+        elif self.handle_missing == 'ignore':
+            for var in self.variables_:
+                self.encoder_dict_[var] = (
+                    X[var]
+                    .value_counts(dropna=True)
+                    .head(self.top_categories)
+                    .index
+                    .tolist()
+                )
+
+        self._check_encoding_dictionary()
+
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        """
+        Replaces the categorical variables with the similarity variables.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The data to transform.
+
+        Returns
+        -------
+        X_new: pandas dataframe.
+            The transformed dataframe. The shape of the dataframe will be different from
+            the original as it includes the similarity variables in place of the of the
+            original categorical ones.
+        """
+
+        check_is_fitted(self)
+        X = self._check_transform_input_and_state(X)
+        if self.handle_missing == 'error':
+            _check_contains_na(X, self.variables_)
+
+        for var in self.variables_:
+            if self.handle_missing == 'impute':
+                X[var] = X[var].fillna('')
+            new_categories = X[var].dropna().unique()
+            column_encoder_dict = {
+                x: _gpm_fast_vec(x, self.encoder_dict_[var]) for x in new_categories
+            }
+            column_encoder_dict[np.nan] = [np.nan] * len(self.encoder_dict_[var])
+            encoded = np.vstack(
+                X[var].map(column_encoder_dict).values
+            )
+            new_features = np.asarray(
+                [f'{var}_{i}' if i else f'{var}_None' for i in self.encoder_dict_[var]]
+            )
+            X.loc[:, new_features] = encoded
+            if self.handle_missing == 'ignore':
+                X.loc[X[var].isna(), new_features] = np.nan
+
+        return X.drop(self.variables_, axis=1)
+
+    def get_feature_names_out(self, input_features: Optional[List] = None) -> List:
```

Comment:
We've got a Mixin for this method that handles all the checks and makes it compatible with the sklearn pipeline.
https://github.com/feature-engine/feature_engine/blob/main/feature_engine/_base_transformers/mixins.py

We only need to add 2 functions, 1 to generate the new feature names, and the second one that would be used by the mixin to add the new features, see for example the OneHotEncoder

https://github.com/feature-engine/feature_engine/blob/93cb37b6650f65281c44ec6d8555d3fb0e55a82c/feature_engine/encoding/one_hot.py#L267-L281

Also, if we have the function to create the new feature names, we don't have to repeat the logic in this method and then also in transform. We can call this function from both. Would you be able to adjust this logic?